package/upgrade_tool/src/mtd.c

```diff
@@ -0,0 +1,281 @@
+                         uint32_t offset, const void *data, uint32_t length)
+{
+  /* verify params */
+  if (mtd_params_verify(partition_info, offset, length, MTD_OP_WRITE) != 0) {
+    return -1;
+  }
+
+  /* open MTD */
+  int fd = mtd_open(partition_info->mtd_num);
+  if (fd < 0) {
+    return -1;
+  }
+
+  /* seek to offset */
+  if (lseek(fd, offset, SEEK_SET) != offset) {
+    printf("error seeking flash\n");
+    return -1;
+  }
+
+  debug_printf("writing mtd%d 0x%08x - 0x%08x...\n",
+               partition_info->mtd_num, offset, offset + length);
+
+  /* write data */
+  if (write(fd, (const void *)data, length) != length) {
+    printf("error writing flash\n");
+    return -1;
+  }
+
+  /* seek to offset */
+  if (lseek(fd, offset, SEEK_SET) != offset) {
+    printf("error seeking flash\n");
+    return -1;
+  }
+
+  /* allocate read buffer */
+  void *buffer = malloc(BUFFER_SIZE);
+  if (buffer == NULL) {
+    printf("error allocating buffer\n");
+    return -1;
+  }
+
+  /* verify data */
+  uint32_t bytes_read = 0;
+  while (bytes_read < length) {
+    uint32_t read_length = length - bytes_read;
+    if (read_length > BUFFER_SIZE) {
+      read_length = BUFFER_SIZE;
+    }
+
+    /* read data */
+    if (read(fd, buffer, read_length) != read_length) {
+      printf("error reading flash\n");
+      return -1;
+    }
+
+    /* compare with source data */
+    if (memcmp(buffer, &((const uint8_t *)data)[bytes_read],
+               read_length) != 0) {
+      printf("error verifying flash\n");
+      return -1;
+    }
+
+    bytes_read += read_length;
+  }
+
+  debug_printf("ok\n");
+
+  /* free buffer */
+  free(buffer);
+
+  /* close MTD */
+  mtd_close(fd);
+  return 0;
+}
+
+int mtd_read(const partition_info_t *partition_info,
+             uint32_t offset, void *buffer, uint32_t length)
+{
+  /* verify params */
+  if (mtd_params_verify(partition_info, offset, length, MTD_OP_READ) != 0) {
+    return -1;
+  }
+
+  /* open MTD */
+  int fd = mtd_open(partition_info->mtd_num);
+  if (fd < 0) {
+    return -1;
+  }
+
+  /* seek to offset */
+  if (lseek(fd, offset, SEEK_SET) != offset) {
+    printf("error seeking flash\n");
+    return -1;
+  }
+
+  /* read data */
+  if (read(fd, buffer, length) != length) {
+    printf("error reading flash\n");
+    return -1;
+  }
+
+  /* close MTD */
+  mtd_close(fd);
+  return 0;
+}
+
+int mtd_crc_compute(const partition_info_t *partition_info,
+                    uint32_t offset, uint32_t length, uint32_t *crc)
+{
+  /* verify params */
+  if (mtd_params_verify(partition_info, offset, length, MTD_OP_READ) != 0) {
+    return -1;
+  }
+
+  /* open MTD */
+  int fd = mtd_open(partition_info->mtd_num);
+  if (fd < 0) {
+    return -1;
+  }
+
+  /* seek to offset */
+  if (lseek(fd, offset, SEEK_SET) != offset) {
+    printf("error seeking flash\n");
+    return -1;
+  }
+
+  /* allocate read buffer */
+  void *buffer = malloc(BUFFER_SIZE);
```

Comment:
Could, but it's already 64k and could potentially be larger. Do you see any issues with malloc? Or what would you have done?